src/Microsoft.PowerShell.ConsoleHost/host/msh/Executor.cs

```diff
@@ -320,11 +322,346 @@ internal Collection<PSObject> ExecuteCommand(string command, out Exception excep
+                (_parent.RunspaceRef.Runspace.ExecutionContext.Modules != null && !_parent.RunspaceRef.Runspace.ExecutionContext.Modules.IsImplicitRemotingModuleLoaded))
+            {
+                return false;
+            }
+
+            try
+            {
+                var scriptBlock = ScriptBlock.Create(command);
+                var scriptBlockAst = scriptBlock.Ast as ScriptBlockAst;
+                if (scriptBlockAst == null)
+                {
+                    throw new ImplicitRemotingBatchingNotSupportedException(ConsoleHostStrings.ImplicitBatchInvalidScriptBlock);
+                }
+
+                // Make sure that this is a simple pipeline
+                string errorId;
+                string errorMsg;
+                scriptBlockAst.GetSimplePipeline(true, out errorId, out errorMsg);
+                if (errorId != null)
+                {
+                    throw new ImplicitRemotingBatchingNotSupportedException(ConsoleHostStrings.ImplicitBatchScriptNotSimplePipeline);
+                }
+
+                // Run checker
+                var checker = new PipelineForBatchingChecker { ScriptBeingConverted = scriptBlockAst };
+                scriptBlockAst.InternalVisit(checker);
+
+                // If this is just a single command, there is no point in batching it
+                if (checker.Commands.Count < 2)
+                {
+                    throw new ImplicitRemotingBatchingNotSupportedException(ConsoleHostStrings.ImplicitBatchSingleCommand);
+                }
+
+                // We have a valid batching candidate
+                using (var ps = System.Management.Automation.PowerShell.Create())
+                {
+                    ps.Runspace = _parent.RunspaceRef.Runspace;
+
+                    // Check commands
+                    ps.Commands.AddCommand("Get-Command").AddParameter("Name", checker.Commands.ToArray());
+                    var cmdInfoList = ps.Invoke<CommandInfo>();
+                    if (ps.Streams.Error.Count > 0)
+                    {
+                        // Cannot continue
+                        throw new PSInvalidOperationException();
+                    }
+                    // All command modules must be implicit remoting modules from the same PSSession
+                    var success = true;
+                    var psSessionId = Guid.Empty;
+                    foreach (var cmdInfo in cmdInfoList)
+                    {
+                        // Check for allowed command
+                        var aliasInfo = cmdInfo as AliasInfo;
+                        string cmdName = (aliasInfo != null) ? aliasInfo.ReferencedCommand.Name : cmdInfo.Name;
+                        if (AllowedCommands.Contains(cmdName))
+                        {
+                            continue;
+                        }
+
+                        if (cmdInfo.Module == null || string.IsNullOrEmpty(cmdInfo.ModuleName))
```

Comment:
== null and string.IsNullOrEmpty check seems redundant.

Reply:
Actually Module and ModuleName properties seem redundant to me.  I am checking both properties to be safe.

Reply:
Ah, misread it as ModuleName twice.